engine/kv_engine_cleaner.cpp

```diff
@@ -0,0 +1,466 @@
+                  }
+                }
+                break;
+              }
+              case PointerType::List: {
+                List* list = slot_iter->GetIndex().list;
+                if (list->GetExpireTime() <= now &&
+                    list->GetTimeStamp() < min_snapshot_ts) {
+                  hash_table_->Erase(&(*slot_iter));
+                  outdated_lists.emplace_back(std::make_pair(
+                      version_controller_.GetCurrentTimestamp(), list));
+                }
+                break;
+              }
+              case PointerType::HashList: {
+                HashList* hlist = slot_iter->GetIndex().hlist;
+                if (hlist->GetExpireTime() <= now &&
+                    hlist->GetTimeStamp() < min_snapshot_ts) {
+                  outdated_hash_lists.emplace_back(std::make_pair(
+                      version_controller_.GetCurrentTimestamp(), hlist));
+                  hash_table_->Erase(&(*slot_iter));
+                }
+                break;
+              }
+              default:
+                break;
+            }
+          }
+          slot_iter++;
+        }
+        hashtable_iter.Next();
+      }  // Finish a slot.
+
+      auto new_ts = version_controller_.GetCurrentTimestamp();
+
+      if (old_string_records.size() > kMaxCachedOldRecords) {
+        pending_purge_strings.emplace_back(
+            PendingPurgeOldRecords{old_string_records, new_ts});
+        old_string_records.clear();
+      }
+
+      if (old_dl_records.size() > kMaxCachedOldRecords) {
+        pending_purge_dls.emplace_back(
+            PendingPurgeOldRecords{old_dl_records, new_ts});
+        old_dl_records.clear();
+      }
+
+      if (outdated_records.size() > kMaxCachedOldRecords) {
+        pending_purge_outdated.emplace_back(
+            PendingPurgeOutDatedRecords{outdated_records, new_ts});
+        outdated_records.clear();
+      }
+
+      {  // purge pending old string records
+        auto iter = pending_purge_strings.begin();
+        while (iter != pending_purge_strings.end()) {
+          if (iter->release_time < version_controller_.OldestSnapshotTS()) {
+            destroyOldStringRecords(iter->records);
+            iter++;
+          } else {
+            break;
+          }
+        }
+        pending_purge_strings.erase(pending_purge_strings.begin(), iter);
+      }
+
+      {  // purge pending old dl records
+        auto iter = pending_purge_dls.begin();
+        while (iter != pending_purge_dls.end()) {
+          if (iter->release_time < version_controller_.OldestSnapshotTS()) {
+            destroyOldDLRecords(iter->records);
+            iter++;
+          } else {
+            break;
+          }
+        }
+        pending_purge_dls.erase(pending_purge_dls.begin(), iter);
+      }
+
+      {  // purge pending outdated records
+        auto outdated_iter = pending_purge_outdated.begin();
+        while (outdated_iter != pending_purge_outdated.end()) {
```

Comment:
Use for loop and put terminate condition together if possible.
Since this is a deque, you can use front and pop_front for convenience.